scanpy/metrics/_morans_i.py

```diff
@@ -0,0 +1,263 @@
+        vals = _get_obs_rep(adata, use_raw=use_raw, layer=layer, obsm=obsm, obsp=obsp).T
+    return morans_i(g, vals)
+
+
+@njit(
+    ft(ft[:], it[:], it[:], ft[:], ft[:], ft, ft),
+    cache=True,
+    parallel=False,
+    fastmath=True,
+)
+def _morans_i_vec_W(
+    x: np.ndarray,
+    indptr: np.ndarray,
+    indices: np.ndarray,
+    data: np.ndarray,
+    z: np.ndarray,
+    W: fp,
+    z2ss: fp,
+) -> Any:
+
+    zl = np.empty(len(z))
+    N = len(indptr) - 1
+
+    for i in prange(N):
+        s = slice(indptr[i], indptr[i + 1])
+        i_indices = indices[s]
+        i_data = data[s]
+        zl[i] = np.sum(i_data * z[i_indices])
+    inum = (z * zl).sum()
+
+    return len(x) / W * inum / z2ss
+
+
+@njit(ft(ft[:], it[:], it[:], ft[:]), cache=True, parallel=True, fastmath=True)
+def _morans_i_vec(
+    x: np.ndarray,
+    indptr: np.ndarray,
+    indices: np.ndarray,
+    data: np.ndarray,
+) -> Any:
+    W = data.sum()
+    z = x - x.mean()
+    z2ss = (z * z).sum()
+    return _morans_i_vec_W(x, indptr, indices, data, z, W, z2ss)
+
+
+@njit(ft[:](ft[:, :], it[:], it[:], ft[:]), cache=True, parallel=True, fastmath=True)
+def _morans_i_mtx(
+    X: np.ndarray,
+    indptr: np.ndarray,
+    indices: np.ndarray,
+    data: np.ndarray,
+) -> Any:
+    M, N = X.shape
+    assert N == len(indptr) - 1
+    W = data.sum()
+    out = np.zeros(M, dtype=ft)
+    for k in prange(M):
+        x = X[k, :].astype(ft)
+        z = x - x.mean()
+        z2ss = (z * z).sum()
+        out[k] = _morans_i_vec_W(x, indptr, indices, data, z, W, z2ss)
+    return out
+
+
+@njit(
+    ft[:](ft[:], it[:], it[:], it[:], it[:], ft[:], tt(it, 2)),
+    cache=True,
+    parallel=True,
+    fastmath=True,
+)
+def _morans_i_mtx_csr(
+    X_data: np.ndarray,
+    X_indices: np.ndarray,
+    X_indptr: np.ndarray,
+    indices: np.ndarray,
+    indptr: np.ndarray,
+    data: np.ndarray,
+    X_shape: tuple,
+):
+    M, N = X_shape
+    W = data.sum()
+    out = np.zeros(M, dtype=ft)
+    x_data_list = np.split(X_data, X_indptr[1:-1])
+    x_indices_list = np.split(X_indices, X_indptr[1:-1])
+    for k in prange(M):
+        x = np.zeros(N, dtype=ft)
+        x_index = x_indices_list[k]
+        x[x_index] = x_data_list[k]
+        z = x - x.mean()
+        z2ss = (z * z).sum()
+        out[k] = _morans_i_vec_W(x, indptr, indices, data, z, W, z2ss)
+    return out
+
+
+###############################################################################
+# Interface (taken from gearys C)
+###############################################################################
+@singledispatch
+def _resolve_vals(val):
+    return np.asarray(val)
+
+
+@_resolve_vals.register(np.ndarray)
+@_resolve_vals.register(sparse.csr_matrix)
+def _(val):
+    return val
+
+
+@_resolve_vals.register(sparse.spmatrix)
+def _(val):
+    return sparse.csr_matrix(val)
+
+
+@_resolve_vals.register(pd.DataFrame)
+@_resolve_vals.register(pd.Series)
+def _(val):
+    return val.to_numpy()
```

Comment:
Could this share the function with `_gearys_c`?

Reply:
yes, importing it from `_gearys_c`! very bad copying over from my side